### ISSUE TITLE
Tissue prior output volumes in random order

### ISSUE THREAD

**Issue by tangwei** [dipy/dipy#3213]
## Description
I generated tissue maps from some pre-computed Q-ball peaks:

`from dipy.segment.tissue import TissueClassifierHMRF`
`hmrf = TissueClassifierHMRF()`
`apm = anisotropic_power(qball_peaks.shm_coeff)`
`results = hmrf.classify(apm, nclass=3, beta=.1)`

The third output in results is the PVE volume which contains the 3 identified classes. They follow a random order when I compared the volumes from different subjects using the same script. Is there a way to make the volumes follow a fixed order? Here is what MRTrix used:

```
0: Cortical grey matter
1: Sub-cortical grey matter
2: White matter
3: CSF
4: Pathological tissue
```
https://mrtrix.readthedocs.io/en/dev/quantitative_structural_connectivity/act.html

[If reporting a bug, attach the entire traceback from Python and follow the way to reproduce below]
[If proposing an enhancement/new feature, provide links to related articles, reference examples, etc.]

## Way to reproduce
[If reporting a bug, please include the following important information:]
- [ ] Code example
- [ ] Relevant images (if any)
- [ ] Operating system and version (run `python -c "import platform; print(platform.platform())"`)
- [ ] Python version (run `python -c "import sys; print('Python', sys.version)"`)
- [ ] dipy version (run `python -c "import dipy; print(dipy.__version__)"`)
- [ ] dependency version (numpy, scipy, nibabel, h5py, cvxpy, fury)
    * import numpy; print("NumPy", numpy.__version__)
    * import scipy; print("SciPy", scipy.__version__)
    * import nibabel; print("Nibabel", nibabel.__version__)
    * import h5py; print("H5py", h5py.__version__)
    * import cvxpy; print("Cvxpy", cvxpy.__version__)
    * import fury; print("fury", fury.__version__)

**Comment by pjsjongsung** [comment]
From what I understand it should not be random, given the innate intensity difference between CSF, GM and WM. Do you have some images of how it is random? For example does CSF and GM change classes with each other sometimes?

**Comment by tangwei** [comment]
From what I see in HCP 7T data, it's random among all three, not only between CSF & GM. Here's an example from two HCP subjects.

Subject 100610:

0:
<img width="245" alt="image" src="https://github.com/dipy/dipy/assets/147845/09c5a8fb-02a0-484a-ab5d-e9e250c78f6a">

1:
<img width="248" alt="image" src="https://github.com/dipy/dipy/assets/147845/b43f4a1f-bbe8-4453-b1b0-22f32da6f8ba">

2 (basically empty):
<img width="211" alt="image" src="https://github.com/dipy/dipy/assets/147845/a8d71847-38c3-4cfa-86ac-b0cb74b71616">


Subject 102816:

0:
<img width="230" alt="image" src="https://github.com/dipy/dipy/assets/147845/2d1df7d2-f1bc-4fee-8515-b4ec8b4c579a">

1:
<img width="241" alt="image" src="https://github.com/dipy/dipy/assets/147845/14525252-c14f-4862-9a17-5852622ef3bf">

2:
<img width="228" alt="image" src="https://github.com/dipy/dipy/assets/147845/a6239cd0-627c-4edd-a96e-eea2bc3853cc">

**Comment by pjsjongsung** [comment]
I see. I still think it is ordered intensity wise, but it is giving unstable results. You can see that on subject 100610 GM and CSF share labels, which is why WM got label 1 and there is almost nothing for label 2.

I am not sure what happened on 102816, but it seems to be a similar problem of GM and CSF being mixed up.

This seems to be a limit of the method itself, which hopefully we will be able to address later on.
For now, could you try maybe changing the number of classes to 4? I found that using 4 can sometimes give better results (where 0 is CSF, 1 is GM, 2 and 3 is WM)

**Comment by tangwei** [comment]
This would not be a problem with analyses using just a few subjects because one can manually identify the tissue maps. It becomes a real problem with batch processing. For example, I need to run the same algorithm through 100+ subjects and it won't be practical to check each individual. In the latter case, I don't think changing the number of classes would guarantee the consistency across subjects.

**Comment by pjsjongsung** [comment]
I agree. It was just a workaround I found when dealing with few subjects. We could definitely benefit from a more robust/less tweaking required tissue segmentation. But maybe others have an easier solution for now.

**Comment by tangwei** [comment]
Yeah, I have a temporary solution using the aparc+aseg files provided by HCP to match the tissue masks. But general users would benefit more from something independent of additional check files.

**Comment by Garyfallidis** [comment]
I think when we say 3 it is actually 4. 3 + background.

**Comment by Garyfallidis** [comment]
The mean intensity for each class should be different and easy to sort.
Background should have the lowest, csf second lowest, gm should be the third intensity and wm should be the last.
You can use the mean intensity for each class to correct the order of the labels.
To move to a 5 class representation that separates cortical to subcortical wm should not be a problem. But currently it is not available with this HMRF application. But if the current order is random we should fix that.

**Comment by pjsjongsung** [comment]
@Garyfallidis The current order is not random. It appears random because CSF and GM had some issues separating but as you can see from the images WM is always a higher class.

**Comment by Garyfallidis** [comment]
Okay, indeed the errors seems different @pjsjongsung

**Comment by tangwei** [comment]
I see, now that makes sense to me. It's a tissue separation issue, not order issue. Nonetheless the issue seems prominent. I checked the 176 HCP 7T subjects, only about 80 had correct tissue classification like the second subject above. The rest of the subjects mostly had GM + CSF mixing like 100610, but there were 20+ had WM + CSF mixing as well:

0:
<img width="229" alt="image" src="https://github.com/dipy/dipy/assets/147845/238988d0-3a7f-49c7-8746-9cccf4d59385">

1:
<img width="234" alt="image" src="https://github.com/dipy/dipy/assets/147845/fc48980b-9dc5-408a-9c78-92118039ca06">

2:
<img width="200" alt="image" src="https://github.com/dipy/dipy/assets/147845/87abfcdc-9c29-42db-94c5-7694a30bb328">

ACT breaks down with either type of mixing.

**Comment by Garyfallidis** [comment]
Great feedback. Please send us a couple of the data that fail.

**Comment by tangwei** [comment]
Yes, I'm using the HCP 7T data on UITS' storage /N/project/hcp_dcwan. Here are some problematic subjects:

1. GM + CSF mixing: 100610, 406836, 901442
2. WM + CSF mixing: 436845, 200614
3. A weird GM + WM mixing: 111312
